src/libraries/System.Linq.AsyncEnumerable/src/System/Linq/Range.cs

```diff
@@ -35,5 +36,51 @@ static async IAsyncEnumerable<int> Impl(int start, int count)
                 }
             }
         }
+
+#if NET7_0_OR_GREATER
+        /// <summary>Generates a sequence of integral numbers within a specified range.</summary>
+        /// <typeparam name="T">The <see cref="IBinaryInteger{TSelf}"/> type of the elements in the sequence.</typeparam>
+        /// <param name="start">The value of the first <see cref="IBinaryInteger{TSelf}"/> in the sequence.</param>
+        /// <param name="count">The number of sequential <see cref="IBinaryInteger{TSelf}"/> to generate.</param>
+        /// <returns>An <see cref="IAsyncEnumerable{T}"/> that contains a range of sequential integral numbers.</returns>
+        /// <exception cref="ArgumentOutOfRangeException"><paramref name="count"/> is less than 0</exception>
+        /// <exception cref="ArgumentOutOfRangeException"><paramref name="start"/> + <paramref name="count"/> -1 is larger than <see cref="IMinMaxValue{TSelf}.MaxValue"/>.</exception>
+        public static IAsyncEnumerable<T> Range<T>(T start, int count) where T : IBinaryInteger<T>
+        {
+            if (count == 0)
+            {
+                return Empty<T>();
+            }
+
+            if (count < 0)
+            {
+                ThrowHelper.ThrowArgumentOutOfRangeException(nameof(count));
+            }
+
+            T max = start + T.CreateTruncating(count - 1);
+            if (start > max || StartMaxCount(start, max) + 1 != count)
+            {
+                ThrowHelper.ThrowArgumentOutOfRangeException(nameof(count));
+            }
+
+            return Impl(start, count);
+
+            static async IAsyncEnumerable<T> Impl(T start, int count)
+            {
+                for (int i = 0; i < count; i++, start++)
+                {
+                    yield return start;
+                }
```

Comment:
@tannergooding, is this the right implementation? The above non-generic does:
```C#
for (int i = 0; i < count; i++)
{
    yield return start + i;
}
```
Is this proposed implementation better or worse than:
```C#
for (int i = 0; i < count; i++)
{
    yield return start + T.CreateTruncating(i);
}
```
?  I'm not sure what requirements we place on IBinaryInteger, in terms of guarantees we make about ++ actually incrementing the value (as opposed to, say, a float + 1 which might not actually increment).